at_client/lib/src/service/notification_service.dart

```diff
@@ -0,0 +1,135 @@
+import 'dart:async';
+
+import 'package:at_client/src/service/notification_service_impl.dart';
+import 'package:at_commons/at_commons.dart';
+
+abstract class NotificationService {
+  // Gives back stream of notifications from the server to the subscribing client.
+  // Optionally pass a regex to filter notification keys matching the regex.
+  Stream<AtNotification> subscribe({String? regex});
+
+  /// Sends notification to [notificationParams.atKey.sharedWith] atSign.
+  ///
+  /// When await on the method returns [NotificationResult].
+  /// When run asynchronously, register to onSuccess and onError callbacks to get [NotificationResult].
+  ///
+  /// OnSuccess is called when the notification has been delivered to the recipient successfully.
+  ///
+  /// onError is called when the notification could not delivered
+  ///
+  ///* Returns [LateInitializationError] when [NotificationParams.atKey] is not initialized
+  ///* Returns [AtKeyException] when invalid [NotificationParams.atKey.key] is formed or when
+  ///invalid metadata is provided.
+  ///* Returns [InvalidAtSignException] on invalid [NotificationParams.atKey.sharedWith] or [NotificationParams.atKey.sharedBy]
+  ///* Returns [AtClientException] when keys to encrypt the data are not found.
+  ///* Returns [AtClientException] when [notificationParams.notifier] is null when [notificationParams.strategy] is set to latest.
+  ///* Returns [AtClientException] when fails to connect to cloud secondary server.
+  ///
+  /// Usage
+  ///
+  /// ```dart
+  /// var currentAtSign = '@alice'
+  /// ```
+  ///
+  /// 1. To notify update of a key to @bob.
+  ///```dart
+  ///  var key = AtKey()
+  ///    ..key = 'phone'
+  ///    ..sharedWith = '@bob';
+  ///
+  ///  var notification = NotificationServiceImpl(atClient!);
+  /// await notification.notify(NotificationParams.forUpdate(key));
+  ///```
+  ///2. To notify and cache a key to @bob
+  ///```dart
+  ///  var metaData = Metadata()..ttr = '600000';
+  ///  var key = AtKey()
+  ///    ..key = 'phone'
+  ///    ..sharedWith = '@bob'
+  ///    ..metadata = metaData;
+  ///
+  ///  var value = '+1 998 999 4940'
+  ///
+  ///  var notification = NotificationServiceImpl(atClient!);
+  /// await notification.notify(NotificationParams.forUpdate(key, value: value));
+  ///```
+  ///3. To notify deletion of a key to @bob.
+  ///```dart
+  ///  var key = AtKey()
+  ///     ..key = 'phone'
+  ///     ..sharedWith = '@bob';
+  ///
+  ///   var notification = NotificationServiceImpl(atClient!);
+  ///   await notification.notify(NotificationParams.forDelete(key));
+  ///```
+  ///4. To notify a text message to @bob
+  ///```dart
+  ///   var notification = NotificationServiceImpl(atClient!);
+  ///   await notification.notify(NotificationParams.forText('Hello','@bob'));
+  ///```
+  Future<NotificationResult> notify(NotificationParams notificationParams,
```

Comment:
@kalluriramkumar We need to document the structure of functions and what you would pass if it has to be one argument or two arguments etc...